Support __float128 with GCC on Apple Silicon

tautschnig opened the pull request:
Although not officially documented it seems that GCC running on Apple Silicon accepts __float128 as a keyword.

Fixes: #8305

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Support __float128 with GCC on Apple Silicon

Although not officially documented it seems that GCC running on Apple
Silicon accepts __float128 as a keyword.

Fixes: #8305

kroening commented:
Is it a built-in typedef or a keyword?

kroening commented on src/ansi-c/ansi_c_internal_additions.cpp:
```diff
@@ -259,6 +259,16 @@ void ansi_c_internal_additions(std::string &code, bool support_float16_type)
+      config.ansi_c.arch == "arm64" &&
+      config.ansi_c.os == configt::ansi_ct::ost::OS_MACOS &&
+      config.ansi_c.mode != configt::ansi_ct::flavourt::CLANG &&
+      config.ansi_c.gcc__float128_type)
```
Can `gcc__float128_type` be false when running GCC on ARM64 Macs?

tautschnig replied:
It is quite possible that no such (old) GCC version even builds on ARM64 Macs, so likely: no.

tautschnig commented:
> Is it a built-in typedef or a keyword?

Ah, it indeed appears to be a keyword. I have updated all comments to reflect this (and explain why we still handle it as if it were a typedef).

kroening commented:
> > Is it a built-in typedef or a keyword?
> 
> Ah, it indeed appears to be a keyword. I have updated all comments to reflect this (and explain why we still handle it as if it were a typedef).

I'd be in favour of adding `PARSER.__float128_keyword` or the like.

tautschnig committed:
Be precise as to whether __float128 is a keyword or a typedef

Add a new parser configuration setting to correctly configure the
scanner.

tautschnig commented:
> I'd be in favour of adding `PARSER.__float128_keyword` or the like.

I've added a second commit that changes the approach implement your suggestion.

kroening commented:
There's lots of instances of getting the `clang` version here, which is expensive, and we don't appear to be using the version number of clang.

tautschnig commented:
> There's lots of instances of getting the `clang` version here, which is expensive, and we don't appear to be using the version number of clang.

We don't need the version number of `clang` for this one, but actually this amounts to a bugfix: https://github.com/diffblue/cbmc/blob/develop/src/ansi-c/gcc_version.cpp#L166-L183 actually want the version number.

tautschnig committed:
Clang configuration: apply defaults rather than using configure_gcc

This isn't as precise, but will avoid the cost of process invocation.
Such cost is considerable on Windows and appears to be around 100ms on
macOS.

feliperodri reviewed (approved):
I assume there are regression tests covering these cases.

tautschnig commented:
> I assume there are regression tests covering these cases.

With https://github.com/diffblue/cbmc/pull/8328 they will!

tautschnig commented:
Closing as the commits are now part of #8328.
